# Machine Learning Classification of Alzheimerâ€™s Disease
## Kristen Norray, Henry Hollis, Zeal Jinwala

In this project, we take GEO series microarray datasets, containing gene expression from several tissues (brain and blood), and train a classifier to predict Alzheimer's Disease (AD) state (Control vs Mild Cognitive Impairment MCI or Control vs AD). 
## Current Usage Instructions
>- Download the latest version of R (4.2) from https://mirrors.upm.edu.my/CRAN/, and do not get the arm version,
even if you have an M1 chip (some packages do not work with this version of R).
>- This project was developed to be run on Apple computers with R 4.2, it can still work on Windows but some extra steps MAY be required.
>- Open Project.mlx, all results and figures are generated by this live script.
>- Data downloading and processing is handled by R, on both PC and Mac, Rexe function should automatically find the location to R path. If not, you need to open R and type R.home("bin"). Take this path (including Rscript.exe) and replace the "Rexe()" in Project.mlx. Warning: if your path has spaces in it, such as C:/Program Files/... enclose in double quotes like: C:/"Program Files"/...
>- If you encounter an error installing packages in R (cell 2 of Project.mlx), follow instructions for running install_packages.R
>- All figures generated by this project appear in the bmes.datadir temp directory.



## Files
- **Project.mlx**: This is the main notebook script that generates all the figures and results.
- **get_binary_labels.m**: This function is useful for generating binary labels for particular classes in the data. Since we train a CTL vs MCI classifier and a CTL vs AD classifier, this is useful for filtering out the sample labels of classes not relevant to particular task (AD and MCI, respectively). 
- **perform_cv_fs.m**: This function has two-fold purpose. First it performs feature selection (restricting genes down from 12k), and second cross-validation on the training data.
- **read_in_data_meta.m**: This function takes the GEO numbers of interest and finds the data and meta files that have been processed in R.
- **Rexe.m**: Script to help find the path to R, may not work always, in which case you must manually find R path.
- **GEO_z_norm.R**: R script that calls geoQuery package to download geo series into a temp directory. We also perform z normalization on downloaded data and filter out the genes not in all the geo series. Finally, we return the subject metadata, which is later used for labeling the patients.
- **enrich_script.R**: R script that takes in tables of GO:BP terms enriched in a gene list and outputs a pretty tree picture with the rrvgo library
- **venn_diagram_de_genes.R**: R script that takes in differentially expressed genes table such as output by limma package, and plots venn diagram of genes. 
- **install_packages.R**: This script is only intended to be run if you encounter the R error where you do not have permission to install packages to R/Library. You would see this message in the 2nd cell of the Project.mlx script