{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Querying the Gene Ontology Database\n",
    "\n",
    "In this tutorial, I am looking at several questions that make use of the Gene Ontology (GO) database and build SQL queries to answer these questions. In each case, I work step-by-step to build the final query. When you are designing your own queries, I recommend that you also work step-by-step; but your final work should only contain the final query that you come up with.\n",
    "\n",
    "Gene Ontology Database contains information about biological terms (GO terms). GO terms include biological processes, molecular functions, and cellular localization; such as \"nucleus\", \"apoptosis\", \"positive regulation of inflammatory response\". GO database also provides associations between genes to GO terms. GO database is one of the primary resources that researchers use to look up the functionality of genes, as described by GO terms.\n",
    "\n",
    "The tables used in this tutorial include: species, gene_product (information about genes), term (where GO terms are listed), graph_path (where relationships between terms are listed), association (which acts as a join table between terms and gene_products).\n",
    "\n",
    "GO terms are organized in a hierarchical structure, e.g., the term \"nucleus\" would encompass other terms like \"nuclear chromosome\", \"nuclear membrane\", etc. The relationship between terms is represented in the graph_path table that joins different terms. In order to find all genes associated with the term \"nucleus\", you need to first collect all the other terms encompassed under \"nucleus\" by collecting them from the graph_path table, and then using the association table to join all these terms with the genes that are associated with.\n",
    "\n",
    "\n",
    "The database schema can also be found at: http://www-legacy.geneontology.org/images/diag-godb-er.jpg\n",
    "\n",
    "![GO database ER diagram](https://sacan.biomed.drexel.edu/lib/exe/fetch.php?media=course:bcomp2:db:godb.er.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os; sys.path.append(os.environ['BMESAHMETDIR']); import bmes\n",
    "import sqlite3\n",
    "\n",
    "from pandas import DataFrame\n",
    "def myselect(sql):\n",
    "    cur.execute(sql);\n",
    "    rows=cur.fetchall();\n",
    "    if len(rows)==0:\n",
    "        print('No results returned for SQL query.');\n",
    "    else:\n",
    "        df = DataFrame(rows)\n",
    "        df.columns = [x[0] for x in cur.description]\n",
    "        display(df) #display() is ipython-specific. In non-ipython script, you can use print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "godbfile = bmes.downloadurl('http://sacan.biomed.drexel.edu/ftp/binf/godb.sqlite')\n",
    "conn = sqlite3.connect(godbfile);\n",
    "cur = conn.cursor();\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retrieve the names of the species that are under the genus ‘Drosophila’"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>ncbi_taxa_id</th>\n",
       "      <th>common_name</th>\n",
       "      <th>lineage_string</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>left_value</th>\n",
       "      <th>right_value</th>\n",
       "      <th>taxonomic_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1107</td>\n",
       "      <td>48328</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>sejuncta</td>\n",
       "      <td>1083217</td>\n",
       "      <td>1202877</td>\n",
       "      <td>1202878</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1645</td>\n",
       "      <td>937309</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>poonia</td>\n",
       "      <td>621428</td>\n",
       "      <td>1202658</td>\n",
       "      <td>1202659</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2828</td>\n",
       "      <td>66368</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>guttifera</td>\n",
       "      <td>78683</td>\n",
       "      <td>1201746</td>\n",
       "      <td>1201747</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5255</td>\n",
       "      <td>48369</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>divaricata</td>\n",
       "      <td>127578</td>\n",
       "      <td>1202793</td>\n",
       "      <td>1202794</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5841</td>\n",
       "      <td>937288</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>inciliata</td>\n",
       "      <td>997796</td>\n",
       "      <td>1202499</td>\n",
       "      <td>1202500</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5921</td>\n",
       "      <td>242880</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>guayllabambae</td>\n",
       "      <td>273046</td>\n",
       "      <td>1201789</td>\n",
       "      <td>1201790</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6762</td>\n",
       "      <td>112146</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>bicornuta</td>\n",
       "      <td>199226</td>\n",
       "      <td>1203161</td>\n",
       "      <td>1203162</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8847</td>\n",
       "      <td>381459</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>cf. clefta BCW-2006</td>\n",
       "      <td>1468585</td>\n",
       "      <td>1203535</td>\n",
       "      <td>1203536</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>11490</td>\n",
       "      <td>88884</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>pallidifrons</td>\n",
       "      <td>8889</td>\n",
       "      <td>1202187</td>\n",
       "      <td>1202188</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>18730</td>\n",
       "      <td>40367</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>flavomontana</td>\n",
       "      <td>1369357</td>\n",
       "      <td>1202374</td>\n",
       "      <td>1202375</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id  ncbi_taxa_id common_name lineage_string       genus  \\\n",
       "0   1107         48328        None           None  Drosophila   \n",
       "1   1645        937309        None           None  Drosophila   \n",
       "2   2828         66368        None           None  Drosophila   \n",
       "3   5255         48369        None           None  Drosophila   \n",
       "4   5841        937288        None           None  Drosophila   \n",
       "5   5921        242880        None           None  Drosophila   \n",
       "6   6762        112146        None           None  Drosophila   \n",
       "7   8847        381459        None           None  Drosophila   \n",
       "8  11490         88884        None           None  Drosophila   \n",
       "9  18730         40367        None           None  Drosophila   \n",
       "\n",
       "               species  parent_id  left_value  right_value taxonomic_rank  \n",
       "0             sejuncta    1083217     1202877      1202878           None  \n",
       "1               poonia     621428     1202658      1202659           None  \n",
       "2            guttifera      78683     1201746      1201747           None  \n",
       "3           divaricata     127578     1202793      1202794           None  \n",
       "4            inciliata     997796     1202499      1202500           None  \n",
       "5        guayllabambae     273046     1201789      1201790           None  \n",
       "6            bicornuta     199226     1203161      1203162           None  \n",
       "7  cf. clefta BCW-2006    1468585     1203535      1203536           None  \n",
       "8         pallidifrons       8889     1202187      1202188           None  \n",
       "9         flavomontana    1369357     1202374      1202375           None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT *\n",
    " FROM species\n",
    " WHERE genus=\"Drosophila\"\n",
    " LIMIT 0,10''')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### •Retrieve the genus and species name of the organisms  whose species name has a prefix ‘mel’ (use LIKE function).\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alnicola</td>\n",
       "      <td>melinoides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hyaloptila</td>\n",
       "      <td>melanosoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Xylaria</td>\n",
       "      <td>mellissii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Trichaster</td>\n",
       "      <td>melanocephalus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pomaria</td>\n",
       "      <td>melanosticta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Stomolophus</td>\n",
       "      <td>meleagris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Isoetes</td>\n",
       "      <td>melanospora</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Gladiolus</td>\n",
       "      <td>meliusculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Brucella</td>\n",
       "      <td>melitensis CNGB 1120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Eleocharis</td>\n",
       "      <td>melanocarpa</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         genus               species\n",
       "0     Alnicola            melinoides\n",
       "1   Hyaloptila            melanosoma\n",
       "2      Xylaria             mellissii\n",
       "3   Trichaster        melanocephalus\n",
       "4      Pomaria          melanosticta\n",
       "5  Stomolophus             meleagris\n",
       "6      Isoetes           melanospora\n",
       "7    Gladiolus           meliusculus\n",
       "8     Brucella  melitensis CNGB 1120\n",
       "9   Eleocharis           melanocarpa"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT genus,species  \n",
    "\t FROM species  \n",
    "\t WHERE species LIKE \"mel%\"  \n",
    "\t LIMIT 0,10''')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### •Retrieve the gene symbols (gene_product.symbol) of all Drosophila melanogaster (species.genus, species.species) genes that are annotated to the ‘nucleus’.\n",
    "\n",
    "Below, I progressively build the query, joining with more and more related tables until all the required tables are referenced. The last query below shows the final query and its result.\n",
    "* Join path: term.id -> graph_path.term1_id /term2_id -> association.term_id /gene_product_id -> gene_product.id\n",
    "* Join path: gene_product.species_id -> species.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>dbxref_id</th>\n",
       "      <th>species_id</th>\n",
       "      <th>type_id</th>\n",
       "      <th>full_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>190680</td>\n",
       "      <td>DDB_G0267178_RTE</td>\n",
       "      <td>291581</td>\n",
       "      <td>1333934</td>\n",
       "      <td>45235</td>\n",
       "      <td>DIRS1 ORF2 fragment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>190681</td>\n",
       "      <td>DDB_G0267182_RTE</td>\n",
       "      <td>291583</td>\n",
       "      <td>1333934</td>\n",
       "      <td>45235</td>\n",
       "      <td>DIRS1 ORF2 fragment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>190682</td>\n",
       "      <td>DDB_G0267188_RTE</td>\n",
       "      <td>291584</td>\n",
       "      <td>1333934</td>\n",
       "      <td>45235</td>\n",
       "      <td>DIRS1 ORF2 fragment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>190683</td>\n",
       "      <td>DDB_G0267206_RTE</td>\n",
       "      <td>291585</td>\n",
       "      <td>1333934</td>\n",
       "      <td>45235</td>\n",
       "      <td>DIRS1 ORF2 fragment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>190684</td>\n",
       "      <td>DDB_G0267210_RTE</td>\n",
       "      <td>291586</td>\n",
       "      <td>1333934</td>\n",
       "      <td>45235</td>\n",
       "      <td>DIRS1 ORF2 fragment</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id            symbol  dbxref_id  species_id  type_id  \\\n",
       "0  190680  DDB_G0267178_RTE     291581     1333934    45235   \n",
       "1  190681  DDB_G0267182_RTE     291583     1333934    45235   \n",
       "2  190682  DDB_G0267188_RTE     291584     1333934    45235   \n",
       "3  190683  DDB_G0267206_RTE     291585     1333934    45235   \n",
       "4  190684  DDB_G0267210_RTE     291586     1333934    45235   \n",
       "\n",
       "             full_name  \n",
       "0  DIRS1 ORF2 fragment  \n",
       "1  DIRS1 ORF2 fragment  \n",
       "2  DIRS1 ORF2 fragment  \n",
       "3  DIRS1 ORF2 fragment  \n",
       "4  DIRS1 ORF2 fragment  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT *  \n",
    "\t FROM gene_product  \n",
    "\t WHERE full_name!=\"\"\n",
    "     LIMIT 0,5\n",
    "     ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>dbxref_id</th>\n",
       "      <th>species_id</th>\n",
       "      <th>type_id</th>\n",
       "      <th>full_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>203433</td>\n",
       "      <td>064Ya</td>\n",
       "      <td>315234</td>\n",
       "      <td>882382</td>\n",
       "      <td>45231</td>\n",
       "      <td>064Ya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>203434</td>\n",
       "      <td>10-4</td>\n",
       "      <td>315237</td>\n",
       "      <td>882382</td>\n",
       "      <td>45231</td>\n",
       "      <td>10-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>203435</td>\n",
       "      <td>11</td>\n",
       "      <td>315239</td>\n",
       "      <td>882382</td>\n",
       "      <td>45231</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>203436</td>\n",
       "      <td>128up</td>\n",
       "      <td>315241</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td>upstream of RpIII128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>203437</td>\n",
       "      <td>14-3-3epsilon</td>\n",
       "      <td>315246</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td>14-3-3epsilon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id         symbol  dbxref_id  species_id  type_id             full_name\n",
       "0  203433          064Ya     315234      882382    45231                 064Ya\n",
       "1  203434           10-4     315237      882382    45231                  10-4\n",
       "2  203435             11     315239      882382    45231                    11\n",
       "3  203436          128up     315241      882382    45236  upstream of RpIII128\n",
       "4  203437  14-3-3epsilon     315246      882382    45236         14-3-3epsilon"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT gene_product.*  \n",
    "\t FROM gene_product, species  \n",
    "\t WHERE species.genus=\"Drosophila\" AND species=\"melanogaster\"  \n",
    "\t AND  species.id = gene_product.species_id \n",
    "\t LIMIT 0,5''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>term_type</th>\n",
       "      <th>acc</th>\n",
       "      <th>is_obsolete</th>\n",
       "      <th>is_root</th>\n",
       "      <th>is_relation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>is_a</td>\n",
       "      <td>relationship</td>\n",
       "      <td>is_a</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>consider</td>\n",
       "      <td>metadata</td>\n",
       "      <td>consider</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>replaced_by</td>\n",
       "      <td>metadata</td>\n",
       "      <td>replaced_by</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Grouping classes that can be excluded</td>\n",
       "      <td>subset</td>\n",
       "      <td>goantislim_grouping</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Term not to be used for direct annotation</td>\n",
       "      <td>subset</td>\n",
       "      <td>gocheck_do_not_annotate</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id                                       name     term_type  \\\n",
       "0   1                                       is_a  relationship   \n",
       "1   2                                   consider      metadata   \n",
       "2   3                                replaced_by      metadata   \n",
       "3   4      Grouping classes that can be excluded        subset   \n",
       "4   5  Term not to be used for direct annotation        subset   \n",
       "\n",
       "                       acc  is_obsolete  is_root  is_relation  \n",
       "0                     is_a            0        0            1  \n",
       "1                 consider            0        0            1  \n",
       "2              replaced_by            0        0            1  \n",
       "3      goantislim_grouping            0        0            0  \n",
       "4  gocheck_do_not_annotate            0        0            0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT *  \n",
    "\t FROM term \n",
    "\t LIMIT 0,5''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>term_type</th>\n",
       "      <th>acc</th>\n",
       "      <th>is_obsolete</th>\n",
       "      <th>is_root</th>\n",
       "      <th>is_relation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4673</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     id     name           term_type         acc  is_obsolete  is_root  \\\n",
       "0  4673  nucleus  cellular_component  GO:0005634            0        0   \n",
       "\n",
       "   is_relation  \n",
       "0            0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT *  \n",
    "\t FROM term \n",
    "\t WHERE name=\"nucleus\"  \n",
    "\t LIMIT 0,5''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>term_type</th>\n",
       "      <th>acc</th>\n",
       "      <th>is_obsolete</th>\n",
       "      <th>is_root</th>\n",
       "      <th>is_relation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>181</td>\n",
       "      <td>nuclear ubiquitin ligase complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0000152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>203</td>\n",
       "      <td>nuclear exosome (RNase complex)</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0000176</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>260</td>\n",
       "      <td>nuclear chromosome</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0000228</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>263</td>\n",
       "      <td>condensed nuclear chromosome</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0000794</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>604</td>\n",
       "      <td>condensed nuclear chromosome, centromeric region</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0000780</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id                                              name           term_type  \\\n",
       "0  181                  nuclear ubiquitin ligase complex  cellular_component   \n",
       "1  203                   nuclear exosome (RNase complex)  cellular_component   \n",
       "2  260                                nuclear chromosome  cellular_component   \n",
       "3  263                      condensed nuclear chromosome  cellular_component   \n",
       "4  604  condensed nuclear chromosome, centromeric region  cellular_component   \n",
       "\n",
       "          acc  is_obsolete  is_root  is_relation  \n",
       "0  GO:0000152            0        0            0  \n",
       "1  GO:0000176            0        0            0  \n",
       "2  GO:0000228            0        0            0  \n",
       "3  GO:0000794            0        0            0  \n",
       "4  GO:0000780            0        0            0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT t2.*  \n",
    "\t FROM term AS t1, graph_path, term AS t2 \n",
    "\t WHERE t1.name=\"nucleus\"  \n",
    "\t AND t1.id = graph_path.term1_id  \n",
    "\t AND graph_path.term2_id = t2.id \n",
    "\t LIMIT 0,5''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>term_id</th>\n",
       "      <th>gene_product_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>631718</td>\n",
       "      <td>181</td>\n",
       "      <td>147756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>983471</td>\n",
       "      <td>181</td>\n",
       "      <td>204446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>987029</td>\n",
       "      <td>181</td>\n",
       "      <td>205622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1004874</td>\n",
       "      <td>181</td>\n",
       "      <td>210318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1004896</td>\n",
       "      <td>181</td>\n",
       "      <td>210319</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id  term_id  gene_product_id\n",
       "0   631718      181           147756\n",
       "1   983471      181           204446\n",
       "2   987029      181           205622\n",
       "3  1004874      181           210318\n",
       "4  1004896      181           210319"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT association.id, association.term_id, association.gene_product_id \n",
    "\t FROM term AS t1, graph_path, term AS t2, association \n",
    "\t WHERE t1.name=\"nucleus\"  \n",
    "\t AND t1.id = graph_path.term1_id  \n",
    "\t AND graph_path.term2_id = t2.id \n",
    "\t AND t2.id = association.term_id  \n",
    "\t LIMIT 0,5''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>dbxref_id</th>\n",
       "      <th>species_id</th>\n",
       "      <th>type_id</th>\n",
       "      <th>full_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>147756</td>\n",
       "      <td>NOT4</td>\n",
       "      <td>247071</td>\n",
       "      <td>95231</td>\n",
       "      <td>45231</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>204446</td>\n",
       "      <td>CG11261</td>\n",
       "      <td>318046</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>205622</td>\n",
       "      <td>CG15800</td>\n",
       "      <td>319580</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>210318</td>\n",
       "      <td>Cul1</td>\n",
       "      <td>318048</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td>Cullin 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>210319</td>\n",
       "      <td>Cul2</td>\n",
       "      <td>326543</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td>Cullin 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id   symbol  dbxref_id  species_id  type_id full_name\n",
       "0  147756     NOT4     247071       95231    45231          \n",
       "1  204446  CG11261     318046      882382    45236          \n",
       "2  205622  CG15800     319580      882382    45236          \n",
       "3  210318     Cul1     318048      882382    45236  Cullin 1\n",
       "4  210319     Cul2     326543      882382    45236  Cullin 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT gene_product.*  \n",
    "\t FROM term AS t1, graph_path, term AS t2, association, gene_product \n",
    "\t WHERE t1.name=\"nucleus\"  \n",
    "\t AND t1.id = graph_path.term1_id  \n",
    "\t AND graph_path.term2_id = t2.id \n",
    "\t AND t2.id = association.term_id  \n",
    "\t AND association.gene_product_id = gene_product.id  \n",
    "\t LIMIT 0,5''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>dbxref_id</th>\n",
       "      <th>species_id</th>\n",
       "      <th>type_id</th>\n",
       "      <th>full_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>203437</td>\n",
       "      <td>14-3-3epsilon</td>\n",
       "      <td>315246</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td>14-3-3epsilon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>203437</td>\n",
       "      <td>14-3-3epsilon</td>\n",
       "      <td>315246</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td>14-3-3epsilon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>203437</td>\n",
       "      <td>14-3-3epsilon</td>\n",
       "      <td>315246</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td>14-3-3epsilon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>203438</td>\n",
       "      <td>14-3-3zeta</td>\n",
       "      <td>315264</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td>14-3-3zeta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>203438</td>\n",
       "      <td>14-3-3zeta</td>\n",
       "      <td>315264</td>\n",
       "      <td>882382</td>\n",
       "      <td>45236</td>\n",
       "      <td>14-3-3zeta</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id         symbol  dbxref_id  species_id  type_id      full_name\n",
       "0  203437  14-3-3epsilon     315246      882382    45236  14-3-3epsilon\n",
       "1  203437  14-3-3epsilon     315246      882382    45236  14-3-3epsilon\n",
       "2  203437  14-3-3epsilon     315246      882382    45236  14-3-3epsilon\n",
       "3  203438     14-3-3zeta     315264      882382    45236     14-3-3zeta\n",
       "4  203438     14-3-3zeta     315264      882382    45236     14-3-3zeta"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myselect('''SELECT gene_product.*  \n",
    "\t FROM term AS t1, graph_path, term AS t2, association, gene_product, species \n",
    "\t WHERE t1.name=\"nucleus\"  \n",
    "\t AND t1.id = graph_path.term1_id  \n",
    "\t AND graph_path.term2_id = t2.id \n",
    "\t AND t2.id = association.term_id  \n",
    "\t AND association.gene_product_id = gene_product.id  \n",
    "\t AND species.genus=\"Drosophila\" AND species=\"melanogaster\"  \n",
    "\t AND  species.id = gene_product.species_id \n",
    "\t LIMIT 0,5''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes found: [2377]. The first 10 of these genes are listed below:\n",
      "14-3-3epsilon, 14-3-3zeta, 2.1, 33-13, 4E-T, ADD1, AGO1, AGO2, AMPKalpha, AP-1mu\n"
     ]
    }
   ],
   "source": [
    "#This is the final form of the query to answer the question above:\n",
    "cur.execute('''SELECT DISTINCT(gene_product.symbol)\n",
    "\t FROM term AS t1, graph_path, term AS t2, association, gene_product, species \n",
    "\t WHERE t1.name=\"nucleus\"  \n",
    "\t AND t1.id = graph_path.term1_id  \n",
    "\t AND graph_path.term2_id = t2.id \n",
    "\t AND t2.id = association.term_id  \n",
    "\t AND association.gene_product_id = gene_product.id  \n",
    "\t AND species.genus=\"Drosophila\" AND species=\"melanogaster\"  \n",
    "\t AND  species.id = gene_product.species_id ''')\n",
    "\n",
    "rs=cur.fetchall();\n",
    "print('Number of genes found: [%d]. The first 10 of these genes are listed below:'%(len(rs)))\n",
    "print(', '.join([x[0] for x in rs[0:10]]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
