<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<base target="_parent" />
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en"
 lang="en" dir="ltr">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <title>
    course:qsb:rnaseq:hwrnaseq.dbp2:index    [Sacan Bioinformatics Lab]
  </title>
  <meta name="generator" content="DokuWiki" />
<meta name="robots" content="index,follow" />
<meta name="date" content="2022-05-24T17:44:15-0400" />
<meta name="keywords" content="course,qsb,rnaseq,hwrnaseq.dbp2,index" />
<link rel="search" type="application/opensearchdescription+xml" href="/lib/exe/opensearch.php" title="Sacan Bioinformatics Lab" />
<link rel="start" href="/" />
<link rel="contents" href="/index.php?id=course:qsb:rnaseq:hwrnaseq.dbp2:index&amp;do=index" title="Sitemap" />
<link rel="alternate" type="application/rss+xml" title="Recent Changes" href="/feed.php" />
<link rel="alternate" type="application/rss+xml" title="Current Namespace" href="/feed.php?mode=list&amp;ns=course:qsb:rnaseq:hwrnaseq.dbp2" />
<link rel="alternate" type="text/html" title="Plain HTML" href="/index.php?do=export_xhtml&amp;id=course:qsb:rnaseq:hwrnaseq.dbp2:index" />
<link rel="alternate" type="text/plain" title="Wiki Markup" href="/index.php?do=export_raw&amp;id=course:qsb:rnaseq:hwrnaseq.dbp2:index" />
<link rel="canonical" href="https://sacan.biomed.drexel.edu/index.php?id=course:qsb:rnaseq:hwrnaseq.dbp2:index" />
<link rel="stylesheet" media="screen" type="text/css" href="/lib/exe/css.php?t=arctic&amp;tseed=1624228061" />
<link rel="stylesheet" media="all" type="text/css" href="/lib/exe/css.php?s=all&amp;t=arctic&amp;tseed=1624228061" />
<link rel="stylesheet" media="print" type="text/css" href="/lib/exe/css.php?s=print&amp;t=arctic&amp;tseed=1624228061" />
<script type="text/javascript" ><!--//--><![CDATA[//><!--
var NS='course:qsb:rnaseq:hwrnaseq.dbp2';var PURGE=0;var JSINFO = {"id":"course:qsb:rnaseq:hwrnaseq.dbp2:index","namespace":"course:qsb:rnaseq:hwrnaseq.dbp2"};
//--><!]]></script>
<script type="text/javascript" charset="utf-8" src="/lib/exe/js.php?tseed=1624228061&amp;t=arctic" ></script>
<script type="text/javascript" charset="utf-8" src="/lib/plugins/bmed/tjpzoom/tjpzoom.js" ></script>
<script type="text/javascript" charset="utf-8" ><!--//--><![CDATA[//><!--
addInitEvent(function() { return(decryptEditSetup()); });
//--><!]]></script>
<script type="text/javascript" ><!--//--><![CDATA[//><!--

var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
		  
//--><!]]></script>
<script type="text/javascript" ><!--//--><![CDATA[//><!--

var pageTracker = _gat._getTracker("UA-16490104-1");
pageTracker._initData();
pageTracker._trackPageview();
		  
//--><!]]></script>
<script type="text/javascript" charset="utf-8" src="/lib/plugins/indexmenu/jsmenu/menu.js" ></script>
<script type="text/javascript" charset="utf-8" src="/lib/plugins/indexmenu/indexmenu.js" ></script>
<script type="text/javascript" charset="utf-8" ><!--//--><![CDATA[//><!--
var indexmenu_ID='course:qsb:rnaseq:hwrnaseq.dbp2:index'
//--><!]]></script>
  <link rel="shortcut icon" href="/lib/plugins/arctic/images/favicon.ico" />
  	<style>
div.dokuwiki div.center_page{
width:98% !important;
padding-left:1%;
}
	</style>
</head>
<body  >
<div id="wrapper">
  <div class="dokuwiki">
        <div class="header">
        <div class="logo">
          <a href="/index.php"><img src="/lib/exe/fetch.php?media=bmed.logo.png" /></a>
        </div>
        <div class="sitetitle">
          <a href="/index.php"  name="dokuwiki__top" accesskey="h">Sacan Bioinformatics Lab<br />Drexel University</a>        </div>
        <div class="headerright">
        <div class="pageicons"><ul> <li><div class="li"><a title="Print view" lass="media" href="#" onclick="javascript:location.href='/index.php?id=course%3Aqsb%3Arnaseq%3Ahwrnaseq.dbp2%3Aindex&printview=1&doprint=1';return false;"><img src="/lib/plugins/arctic//images/tool-print.png"></a></div></li> <li><div class="li"><a title="Refresh this page" lass="media" href="#" onclick="javascript:location.href='/index.php?id=course%3Aqsb%3Arnaseq%3Ahwrnaseq.dbp2%3Aindex&purge=1';return false;"><img src="/lib/plugins/arctic//images/refresh.jpg"></a></div></li> <li><div class="li"><a title="Login" class="media" href="/index.php?id=course%3Aqsb%3Arnaseq%3Ahwrnaseq.dbp2%3Aindex&do=login&sectok=715174e1d1b954c06610038e6fedd55a"><img src="/lib/plugins/arctic/images/tool-login.gif"></a></div></li></ul></div>        </div>
      </div>
        <div class="breadcrumbs">
<span class="bchead">You are here: </span><a href="/index.php?id=index"  title="index">index</a> &raquo; <a href="/index.php?id=course:index"  title="course:index">course</a> &raquo; <a href="/index.php?id=course:qsb:index"  title="course:qsb:index">qsb</a> &raquo; <a href="/index.php?id=course:qsb:rnaseq:index"  title="course:qsb:rnaseq:index">rnaseq</a> &raquo; <a href="/index.php?id=course:qsb:rnaseq:hwrnaseq.dbp2:index"  title="course:qsb:rnaseq:hwrnaseq.dbp2:index">hwrnaseq.dbp2</a>  </div>
              <table width=100% cellspacing=0 cellpadding=0><tr><td>
              <div class="center_page">
                    
<h2 class="sectionedit1"><a name="RNA-Seq Analysis" id="RNA-Seq Analysis" href="#RNA-Seq Analysis">RNA-Seq Analysis</a></h2>
<div class="level2">

<p>
TODO: Make large-file usage extra points and using TPM as part of the alignment.
</p>

<p>
In this assignment, you will re-analyze the RNA-Seq reported in the following paper:
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4224148/" class="urlextern" title="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4224148/"  rel="nofollow">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4224148/</a>

</p>
<ul>
<li class="level1"><div class="li">Find the series entry on GEO.</div>
</li>
</ul>
<ul>
<li class="level1"><div class="li">You need to find out the SRR identifiers for wild type and mutant samples. For this assignment, analyzing one run from each group is sufficient.</div>
</li>
</ul>
<ul>
<li class="level1"><div class="li">Download (in your code, if the file isn&#039;t downloaded before) the fastq files from the SRA (NCBI Sequence Read Archive)</div>
<ul>
<li class="level1"><div class="li">The fastq files are not available via your browser. You will need the fastq-dump program (available in SRA Toolkit software) to download the fastq files from NCBI. for fastq-dump, see: <a href="https://edwards.sdsu.edu/research/fastq-dump/" class="urlextern" title="https://edwards.sdsu.edu/research/fastq-dump/"  rel="nofollow">https://edwards.sdsu.edu/research/fastq-dump/</a></div>
</li>
<li class="level1"><div class="li">I recommend working on the smaller dataset mentioned below first. When your code is complete, you can then apply it to the original SRA data. Expect download and analysis of the original SRA files to take up to several hours.</div>
</li>
<li class="level1"><div class="li">If you use fastq-dump, run the command line from your code (e.g., “system()” function in Matlab). Only run the command line if you have not downloaded the fastq file before.</div>
</li>
<li class="level1"><div class="li"><strong>Do not place the fastq files in your dropbox folder.</strong></div>
</li>
</ul>
</li>
</ul>
<ul>
<li class="level1"><div class="li">If you are unable to complete the assignment using the original SRA files, you may use an alternative dataset (for a -5% grade penalty). You may download (in your code, if the file isn&#039;t downloaded before) a random down-sampling of the fastq files available at:</div>
</li>
</ul>

<p>
<a href="http://sacan.biomed.drexel.edu/ftp/rnaseq.dbp2/SRR1302790_pass.randsample.fastq" class="urlextern" title="http://sacan.biomed.drexel.edu/ftp/rnaseq.dbp2/SRR1302790_pass.randsample.fastq"  rel="nofollow">http://sacan.biomed.drexel.edu/ftp/rnaseq.dbp2/SRR1302790_pass.randsample.fastq</a> and <a href="http://sacan.biomed.drexel.edu/ftp/rnaseq.dbp2/SRR1302792_pass.randsample.fastq" class="urlextern" title="http://sacan.biomed.drexel.edu/ftp/rnaseq.dbp2/SRR1302792_pass.randsample.fastq"  rel="nofollow">http://sacan.biomed.drexel.edu/ftp/rnaseq.dbp2/SRR1302792_pass.randsample.fastq</a>
</p>
<ul>
<li class="level1"><div class="li"><strong>Do not place the fastq files in your dropbox folder.</strong></div>
</li>
</ul>
<ul>
<li class="level1"><div class="li">Download (in your code, if the file isn&#039;t downloaded before) the yeast genome fasta file and index it using bwa.</div>
<ul>
<li class="level1"><div class="li">or, download the igenome compilation for yeast, which contains pre-indexed files.</div>
</li>
<li class="level1"><div class="li">Document (as comments in your code) how you got to the genome file(s). e.g., “searched &#039;….&#039; on Google/NCBI, followed NCBI Genome link, followed Genome Fasta file download link, etc.”</div>
</li>
<li class="level1"><div class="li"><strong>Do not place the genome fasta or index files in your dropbox folder.</strong></div>
</li>
</ul>
</li>
</ul>
<ul>
<li class="level1"><div class="li">Map the reads (the fastq files) against the yeast genome file using bmes_bwa(), which is provided to you.</div>
</li>
</ul>
<ul>
<li class="level1"><div class="li">Download (and build/install as needed) the featureCounts program (from the subread package). Download links and documentation available here: <a href="http://subread.sourceforge.net/" class="urlextern" title="http://subread.sourceforge.net/"  rel="nofollow">http://subread.sourceforge.net/</a></div>
<ul>
<li class="level1"><div class="li">You may download it manually (outside your code, from your browser).</div>
</li>
<li class="level1"><div class="li"><strong>Do not place the software or the fastq files in your dropbox folder.</strong></div>
</li>
</ul>
</li>
</ul>
<ul>
<li class="level1"><div class="li">Use featureCounts (from your code, using a system() call) to “enrich” the mapped positions in the sam file into respective gene names.</div>
<ul>
<li class="level1"><div class="li">Don&#039;t do featureCounts of a sam file if it has been done before.</div>
</li>
<li class="level1"><div class="li"><strong>Store the result of featureCounts in your dropbox folder.</strong></div>
</li>
</ul>
</li>
</ul>
<ul>
<li class="level1"><div class="li">Optional: (5% extra credit) Perform TPM normalization of the gene counts data.</div>
</li>
<li class="level0"><div class="li">Compare gene counts between mutant and wild type. List 10 most different (ranked by absolute fold-change) genes between the two groups.</div>
</li>
</ul>
<ul>
<li class="level1"><div class="li">Use the genes that are differentially regulated between mutant and wild type. You have to define/decide what it means to be differentially regulated.</div>
</li>
</ul>
<ul>
<li class="level1"><div class="li">Perform functional enrichment analysis using DAVID. Report the enrichment results using comments or a snapshot image(s) of DAVID enriched annotations list.</div>
</li>
</ul>
<ul>
<li class="level1"><div class="li">Publish your code as a pdf file.</div>
</li>
</ul>

</div>

        </div>
            </td></tr></table>

      <div class="stylefoot">
        <div class="meta">
        <div class="user">
</div>        </div>
      </div>

    		
              <div class="clearer"></div>
              </div>
</div>

</body>
</html>
